### ISSUE TITLE
failed to query event: CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure tensorflow/stream_executor/dnn.cc:613] CUDNN_STATUS_INTERNAL_ERROR

### ISSUE THREAD

**Issue by nectario** [tensorflow/tensorflow#41169]
I am unable to train my model. I get the below error after around the 100th epoch or so (randomly). Sometimes it fails on the 500th or so epoch. There are a few times I won't get this error.

Python Version: 3.7
Tesorflow Version: 2.3
Tensorflow Installed: From binary.
OS: Windows 10
GPU Card: Nvidia Titan V
CUDA: 10.1
CUDNN: 7.6.5.32

: failed to query event: CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure
2020-07-07 17:15:37.551112: E tensorflow/stream_executor/dnn.cc:613] CUDNN_STATUS_INTERNAL_ERROR
in tensorflow/stream_executor/cuda/cuda_dnn.cc(1867): 'cudnnRNNForwardTraining( cudnn.handle(), rnn_desc.handle(), model_dims.max_seq_length, input_desc.handles(), input_data.opaque(), input_h_desc.handle(), input_h_data.opaque(), input_c_desc.handle(), input_c_data.opaque(), rnn_desc.params_handle(), params.opaque(), output_desc.handles(), output_data->opaque(), output_h_desc.handle(), output_h_data->opaque(), output_c_desc.handle(), output_c_data->opaque(), workspace.opaque(), workspace.size(), reserve_space.opaque(), reserve_space.size())'
2020-07-07 17:15:37.551176: F tensorflow/stream_executor/cuda/cuda_dnn.cc:189] Check failed: status == CUDNN_STATUS_SUCCESS (7 vs. 0)Failed to set cuDNN stream.

<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
@nectario,
Could you please provide the complete code to reproduce the issue reported here?

Also, please take a look at [this comment](https://github.com/tensorflow/tensorflow/issues/35950) from a similar issue and let us know if it helps. Thanks!

**Comment by nectario** [comment]
I can't post the code since it's a proprietary model. It seems this is related to NVIDA drivers. This happens RANDOMNLY. Last night it happened on epoch 2500 for example.

**Comment by nectario** [comment]
My model is a 6 layer LSTM.

**Comment by nectario** [comment]
I really do not want to downgrade my NVIDIA GPU in order for this issue to go away. It's becoming impossible to train a model till completion because of this issue.

**Comment by amahendrakar** [comment]
> I can't post the code since it's a proprietary model.

@nectario,
As an alternative, is it possible for you to provide a working example which can mimic the error? Thanks!

**Comment by nectario** [comment]
Yes, I will soon be posting sample code that recreates this issue.

**Comment by nectario** [comment]
This is an extremely serious issue. It prevents me from training a model till the end.

**Comment by nectario** [comment]
Hi,

Here is the code that recreates this issue. I was able to make the error happen twice, although it happens randomly.

First time it happened at epoch **1548** 
Second time it happened at epoch **524**

**To Download:**
[http://www.nektron.com/cudnn_error.zip](url)

**To Run:**
`python python CUDNN_Error.py`

I have **tensorflow 2.3** 
GPU is the **Nvidia Titan V**
I am running this on **Windows 10**

**Note:** you need to have it run for 3000 epochs and it will happen as some point.

**Comment by nectario** [comment]
Please note, in the code you will see a Callback that deliberately tries to load data "on_epoch_end"  just to add a bit of a delay that mimics my own callbacks.

**Comment by amahendrakar** [comment]
@nectario,
I was able to run the code without any issues with TF v2.3. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/0c5687807b21b53df5942c83c4fb5e62/41169.ipynb).

Could you please try setting a hard limit on the total GPU memory as show in [this guide](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth) and check if it works. Thanks!

**Comment by nectario** [comment]
@amahendrakar Thank you. I will try your suggestion. Looking at how you ran it, I don't believe this is repeatable in a server GPU like the V100 or any other ones. It really needs to run in a consumer GPU like the Titan V, and in PC desktop environment. As of now, I can never finish training so it reaches the last epoch. I think it's related to the Nvidia driver.

Also, this gets triggered quicker if I am watching a YouTube video as I am training. I started getting this only after Nvidia updated their drivers.

Here is a person who also experienced a similar issue. Check this link:
[https://forums.developer.nvidia.com/t/crash-on-training-cuda-error-launch-failed/76055](url) 

But it seems it was hardware memory issue which needed a GPU replacement?

**Comment by nectario** [comment]
If someone can run the above in their local desktop environment with the latest Nvidia gpu drivers would be great.

**Comment by nectario** [comment]
I keep getting this issue, in different variation. Very rarely am I able to complete training. It's extremely frustrating...
```
2020-09-01 17:05:07.819801: E tensorflow/stream_executor/dnn.cc:616] CUDNN_STATUS_EXECUTION_FAILED
in tensorflow/stream_executor/cuda/cuda_dnn.cc(1892): 'cudnnRNNForwardTraining( cudnn.handle(), rnn_desc.handle(), model_dims.max_seq_length, input_desc.handles(), input_data.opaque(), input_h_desc.handle(), input_h_data.opaque(), input_c_desc.handle(), input_c_data.opaque(), rnn_desc.params_handle(), params.opaque(), output_desc.handles(), output_data->opaque(), output_h_desc.handle(), output_h_data->opaque(), output_c_desc.handle(), output_c_data->opaque(), workspace.opaque(), workspace.size(), reserve_space.opaque(), reserve_space.size())'
2020-09-01 17:05:07.819965: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure
2020-09-01 17:05:07.861316: W tensorflow/core/framework/op_kernel.cc:1767] OP_REQUIRES failed at cudnn_rnn_ops.cc:1517 : Internal: Failed to call ThenRnnForward with model config: [rnn_mode, rnn_input_mode, rnn_direction_mode]: 2, 0, 0 , [num_layers, input_size, num_units, dir_count, max_seq_length, batch_size, cell_num_units]: [1, 300, 150, 1, 20, 620, 150]
2020-09-01 17:05:07.873534: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:220] Unexpected Event status: 1
```

**Comment by nectario** [comment]
Is this some kind of hardware issue I am facing with my Titan V? Today I was only able to have one successful run.

**Comment by nectario** [comment]
I have made the following observation:
When my room temperature is 73 or below Celsius, the problem does not seem to happen. Generally, on average, my room temperature is 78 Celsius. 
Could this be an overheating issue of the GPU?

**Comment by nectario** [comment]
I think I may have finally resolved this issue. It seems I had two CUDA versions installed: 10.1, and 10.2. My installation was pointing to the 10.2 version. After uninstalling all the other other versions and having only 10.1 I was able to finally train more than once without issues. I will run a couple more tests. Bottom line: After uninstalling all my previous CUDA/CUDNN version and reinstalling the correct ones, it seems to have resolved the issue.

**Comment by argunaykut** [comment]
I had this same issue and read all the comments here. I recognized that a pip install software I made updated my tensorflow to 2.3 which is not even a stable build that conda supports. But my tensorflow-gpu version stayed at 1.14. 
I realized this and forces both my tensorflow and tensorflow-gpu versions to 2.1. This solved my issue.

**Comment by nectario** [comment]
Thank you. Apparently this issue is still not gone for me. It is extremely frustrating. I cannot train till completion. TensorFlow is becoming unusable for me... Even using Python 3.8 with a fresh install I still get this...

**Comment by argunaykut** [comment]
Thank you too, I have been running for 20 minutes without a crash so far. Try Python 3.7 with tensorflow 2.1 and tensorflow-gpu 2.1 :)

**Comment by nectario** [comment]
I hate downgrades, but I will give it a shot!

**Comment by nectario** [comment]
Good luck!

**Comment by ahtik** [comment]
@nectario Confirming, got an error with CUDNN_Error.py at epoch 1683. Python 3.8, TF 2.3.1, CUDA 10.2, RTX 2070S.

Although a bit different error:
```
2020-10-28 11:31:23.136550: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure
2020-10-28 11:31:23.136714: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:220] Unexpected Event status: 1
2020-10-28 11:31:23.136712: E tensorflow/stream_executor/dnn.cc:616] CUDNN_STATUS_EXECUTION_FAILED
in tensorflow/stream_executor/cuda/cuda_dnn.cc(1892): 'cudnnRNNForwardTraining( cudnn.handle(), rnn_desc.handle(), model_dims.max_seq_length, input_desc.handles(), input_data.opaque(), input_h_desc.handle(), input_h_data.opaque(), input_c_desc.handle(), input_c_data.opaque(), rnn_desc.params_handle(), params.opaque(), output_desc.handles(), output_data->opaque(), output_h_desc.handle(), output_h_data->opaque(), output_c_desc.handle(), output_c_data->opaque(), workspace.opaque(), workspace.size(), reserve_space.opaque(), reserve_space.size())'
```
By  the way, switching between CUDA versions in Win is easy without uninstall&reinstall. NVIDIA drivers are backwards compatible with CUDA 10.x and 11.x, so one can always have the latest drivers installed. Just have to make sure to install CUDA & CuDNN separately. nvidia-smi reported CUDA version output is "irrelevant". To switch, I'm using env cmds:
```
SET CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1
SET PATH=%CUDA_PATH%\bin;%PATH%
SET PATH=%CUDA_PATH%\extras\CUPTI\libx64;%PATH%
SET PATH=C:\cudnn-7.6.5\10.1\bin;%PATH%;
```

Training the same example with CUDA 10.1 in progress.

If I get time, will try again the CUDA 10.2 run with `SET CUDA_LAUNCH_BLOCKING=1` to see if there's any extra bits coming out.

**Comment by nectario** [comment]
Wow, yeap, the error message would also change a bit sometimes for me. A few times I am able to complete an entire training run out of pure luck, but 90% of the time it fails. It can fail on epochs 50, 300, 800, 1100 -- it can be anywhere really. 

Let me know how the 10.1 training goes.

Thanks

**Comment by ahtik** [comment]
CUDA 10.1 training succeeds. Might be worth waiting a few weeks for the TF 2.4 release with CUDA 11 support and would be fun to try again. Don't know if it's going to be CUDA 11.0 or 11.1.

To be fair, TensorFlow >=2.1.0 officially supports only CUDA 10.1, in that sense works as expected.

 Btw, my GPU utilization was quite low with this training (batch size was optimal for max gpu memory use), so the error was not coming from power consumption and definitely not from overheating.

**Comment by nectario** [comment]
Thank you. For me it happens with CUDA 10.1. Oh well... At least I now feel confident it is not a hardware issue.

**Comment by ahtik** [comment]
@nectario I did run the same model training on the latest TF 2.4 release candidate (r2.4 branch at 541fd26bd37) with CUDA 11.0 and CuDNN 8.0.4.30.

Unfortunately failed again at epoch 2720. CUDA utilization perfectly fine, ~80%, GPU temperature ~70C.

```
2020-10-29 09:41:11.513784: E tensorflow/stream_executor/dnn.cc:616] CUDNN_STATUS_EXECUTION_FAILED
in tensorflow/stream_executor/cuda/cuda_dnn.cc(1968): 'cudnnRNNBackwardData( cudnn.handle(), rnn_desc.handle(), model_dims.max_seq_length, output_desc.handles(), output_data.opaque(), output_desc.handles(), output_backprop_data.opaque(), output_h_desc.handle(), output_h_backprop_data.opaque(), output_c_desc.handle(), output_c_backprop_data.opaque(), rnn_desc.params_handle(), params.opaque(), input_h_desc.handle(), input_h_data.opaque(), input_c_desc.handle(), input_c_data.opaque(), input_desc.handles(), input_backprop_data->opaque(), input_h_desc.handle(), input_h_backprop_data->opaque(), input_c_desc.handle(), input_c_backprop_data->opaque(), workspace.opaque(), workspace.size(), reserve_space_data->opaque(), reserve_space_data->size())'
2020-10-29 09:41:11.513855: F tensorflow/stream_executor/cuda/cuda_dnn.cc:189] Check failed: status == CUDNN_STATUS_SUCCESS (7 vs. 0)Failed to set cuDNN stream.
2020-10-29 09:41:11.514110: F .\tensorflow/core/kernels/conv_2d_gpu.h:1019] Non-OK-status: GpuLaunchKernel( SwapDimension1And2InTensor3UsingTiles<T, kNumThreads, kTileSize, kTileSize, conjugate>, total_tiles_count, kNumThreads, 0, d.stream(), input, input_dims, output) status: Internal: unspecified launch failure
```

**Comment by nectario** [comment]
I see. It is interesting that your failures occur close to the 3000 epochs. In my case, it almost always happens below 1200.

**Comment by ion-elgreco** [comment]
This really needs to get fixed, LSTM's can't be run with Cuda now.

**Comment by nectario** [comment]
Absolutely. It is even preventing another issue I reported to be debugged because of this issue. It's this one #45676

**Comment by nectario** [comment]
This issue belongs in a range of issues related to LSTMs and CUDA/CUDNN. Each time the error is a bit different but same effect.

**Comment by nectario** [comment]
Probably NVIDIA needs to be involved also for this to be fixed.

**Comment by nectario** [comment]
When I train on AWS on their V100s, this never happens. Seems to be happening mainly with the consumer based graphics cards.

**Comment by ion-elgreco** [comment]
> This issue belongs in a range of issues related to LSTMs and CUDA/CUDNN. Each time the error is a bit different but same effect.

Yeah, I get all of them, sometimes this Cuda error, sometimes the rnnbackward or rnnforward error. The strange thing is that it only occurred after I had swapped out my motherboard and CPU and reinstalled my windows. 

I downloaded the same drivers as before, did a complete reinstall of my conda environment, but somehow I got this issue in the process.

**Comment by nectario** [comment]
I know!! I have been living with this for way too long. More than six months... Do you get this with 2.4?

**Comment by nectario** [comment]
With 2.4 it seems to have made it better, but training got slower.

**Comment by ion-elgreco** [comment]
> I know!! I have been living with this for way too long. More than six months... Do you get this with 2.4?

Yep :( I had bought a couple of months ago a RTX 3080, was using TF 2.4 RC0 up to the final release with no issues. Could run LSTM's without a problem, but after swapping my cpu and motherboard and reinstall windows, I can't seem to get around this problem.

I even reinstalled windows another time, but still getting the issue. So maybe it's also hardware related...

**Comment by nectario** [comment]
Yeah. I suspect it's a combination of gfx driver/tensorflow/cuda/cudnn. Was the gfx driver updated around that time?

**Comment by nectario** [comment]
This issue needs tensorflow heavyweights to look into it! The credibility of the entire framework is in line! I am this close to completely switching frameworks!

**Comment by ion-elgreco** [comment]
> Yeah. I suspect it's a combination of gfx driver/tensorflow/cuda/cudnn. Was the gfx driver updated around that time?

No, I simply reinstalled all the same drivers so that everything was in the same condition.

**Comment by nectario** [comment]
Ooh, I see...

**Comment by nectario** [comment]
It's surely seems there may be some hardware issue because it never happens with the V100s.

**Comment by nectario** [comment]
This requires TensorFlow and NVIDA to look into it...

**Comment by ion-elgreco** [comment]
> It's surely seems there may be some hardware issue because it never happens with the V100s.
Probably a combination of multiple things. It is strange that it did not occur for me before though.

> This requires TensorFlow and NVIDA to look into it...
Definitely, but have they have even acknowledged this issue yet?

**Comment by nectario** [comment]
I believe it's been acknowledged but this issue is unique as it's hard to reproduce. If they try to reproduce this on any cloud machine it will never happen. It requires these consumer based graphics cards. Like I have a Titan V or an RTX 3080 which you have. It's good that we are keeping this thread "warm"

I hope someone from TensorFlow can update us here if they are investigating it on what is causing it.

My resolution for now: I reduced the number of epochs. When it fails, I load the last saved weights and continue on.

**Comment by ion-elgreco** [comment]
> I believe it's been acknowledged but this issue is unique as it's hard to reproduce. If they try to reproduce this on any cloud machine it will never happen. It requires these consumer based graphics cards. Like I have a Titan V or an RTX 3080 which you have. It's good that we are keeping this thread "warm"
> 
> I hope someone from TensorFlow can update us here if they are investigating it on what is causing it.
> 
> My resolution for now: I reduced the number of epochs. When it fails, I load the last saved weights and continue on.

Ah I see, let's hope they are looking into it and they are able to reproduce it. I'll try if just disabling the cudnn lstm kernel is enough. I'll give an update if it works!

**Comment by ion-elgreco** [comment]
@nectario So disabling eager execution helps, because it's not using the cudnn kernel for lstm anymore.

**Comment by ahtik** [comment]
@nectario I'm curious, which cudnn versions did you test with when running with the TF 2.4 final? There's now 8.0.3, 8.0.4.30 and 8.0.5.39, all having CUDA 11.0 support. I presume the same crash still happened with the 2.4 as well, not just the slowdown issue?

**Comment by ion-elgreco** [comment]
> @nectario I'm curious, which cudnn versions did you test with when running with the TF 2.4 final? There's now 8.0.3, 8.0.4.30 and 8.0.5.39, all having CUDA 11.0 support. I presume the same crash still happened with the 2.4 as well, not just the slowdown issue?

I tried all of those versions, but no avail. All of them showed the same behaviour.

**Comment by nectario** [comment]
> @nectario I'm curious, which cudnn versions did you test with when running with the TF 2.4 final? There's now 8.0.3, 8.0.4.30 and 8.0.5.39, all having CUDA 11.0 support. I presume the same crash still happened with the 2.4 as well, not just the slowdown issue?

I initially tested with the latest versions 8.0.5.39 and then went back to the advertised versions.￼

**Comment by ahtik** [comment]
@nectario I ran your cudnn_error.zip through my setup and at least the 1st run worked fine. Although, I did do a couple of optimizations to speed it up.

Started with ~487ms/step, 3s epoch.
Added to the beginning of the CUDNN_Error.py to enable mixed precision and changed all LSTM units from 150 to 152 to be more friendly for the mixed precision (n%8=0)
```
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
```

Went down to ~280ms/step. After enabling mixed precision I could also double the batch size from 620 to 1240 and went further down to ~440ms/step (for previous batch size normalization: 220ms/step). Each epoch down to 1s.

There might be room for XLA-based optimizations, but I'm getting `SubProcess ended with return code: 0` from XLA.

Graphics driver 460.89, CUDA 11.0.2, cudnn-8.0.5.39.

**Comment by nectario** [comment]
@ahtik Thank you! I will try those changes, except the batch change. The batch was carefully selected to yield better metrics. A bigger batch does not necessarily give me better metric performance.

I wonder if this explains the drop in performance in 2.4?

**Comment by nectario** [comment]
> @nectario I ran your cudnn_error.zip through my setup and at least the 1st run worked fine. Although, I did do a couple of optimizations to speed it up.
> 
> Started with ~487ms/step, 3s epoch.
> Added to the beginning of the CUDNN_Error.py to enable mixed precision and changed all LSTM units from 150 to 152 to be more friendly for the mixed precision (n%8=0)
> 
> ```
> from tensorflow.keras.mixed_precision import experimental as mixed_precision
> policy = mixed_precision.Policy('mixed_float16')
> mixed_precision.set_policy(policy)
> ```
> 
> Went down to ~280ms/step. After enabling mixed precision I could also double the batch size from 620 to 1240 and went further down to ~440ms/step (for previous batch size normalization: 220ms/step). Each epoch down to 1s.
> 
> There might be room for XLA-based optimizations, but I'm getting `SubProcess ended with return code: 0` from XLA.
> 
> Win10, Graphics driver 460.89, CUDA 11.0.2, cudnn-8.0.5.39.

@nikitamaia I wonder if the above solution is what is needed to fix the slow performance for #45676?

**Comment by nectario** [comment]
@nikitamaia @ahtik 

Adding your code @ahtik at the start, dropped my per epoch time to 1 second!

from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)

**Comment by ahtik** [comment]
@nectario To be honest, I suspect the 2.3 vs 2.4 slowdown in #45676 difference is still somewhere else because mixed precision hasn't been enabled, so it's an extra improvement on top of whatever else is going on.

**Comment by nectario** [comment]
> @nectario To be honest, I suspect the 2.3 vs 2.4 slowdown in #45676 difference is still somewhere else because mixed precision has never been enabled by default, so it's an extra improvement on top of whatever else is going on.

Got it.

**Comment by ion-elgreco** [comment]
@nectario TF2.3 and Cuda 10.1 and cudnn 7.6.5 don't cause any crashes with the cudnn accelerated LSTMs for me, but the only downside with an RTX 3080 is that the initial loading takes up to 15 min! 👎 So also not really a solution. I even tried to install older versions of windows, but also no success. Something within cudnn 8+ and cuda 11+ is causing this.

**Comment by ahtik** [comment]
@ion-elgreco Could you share your minimim LSTM code that is crashing with 
TF2.4? I was no longer able to reproduce the crash with the setup described 
above with RTX2070S. I did enable mixed precision to speed it up, so maybe 
cudnn didn't get enough stress load... How long it takes in your case to 
crash it?

**Comment by ion-elgreco** [comment]
> @ion-elgreco Could you share your minimim LSTM code that is crashing with TF2.4? I was no longer able to reproduce the crash with the setup described above with RTX2070S. I did enable mixed precision to speed it up, so maybe cudnn didn't get enough stress load... How long it takes in your case to crash it?

def build_base(model_name, units):
    model = Sequential(name=model_name)
    model.add(layers.InputLayer(input_shape=(X_train.shape[1], X_train.shape[2])))
    model.add(layers.LSTM(units, return_sequences=True,))
    model.add(layers.Dropout(0.4))
    model.add(layers.LayerNormalization())
    model.add(layers.Dense(7, activation="softmax", name="Dense_Output"))
    model.compile(
        optimizer="rmsprop",
        loss=CategoricalCrossentropy(),
        metrics=["accuracy"]
    )
    return model

history = NN.fit(
                X_train,
                y_train,
                batch_size=batchsize,
                sample_weight=train_samples_weights[:train_div],
                validation_data=(X_val, y_val),
                epochs=5,
                verbose=1,
                shuffle=True,
            )
`

It will already crash with 5 units after 3 epochs...

**Comment by ahtik** [comment]
@ion-elgreco Can you provide the full code together with batchsize etc hyper parameters, train & val datasets etc. Otherwise, it's a bit too much effort to reproduce.

Does it also crash after adding this to the beginning of the file?
```
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
```

**Comment by ion-elgreco** [comment]
> @ion-elgreco Can you provide the full code together with batchsize etc hyper parameters, train & val datasets etc. Otherwise, it's a bit too much effort to reproduce.
> 
> Does it also crash after adding this to the beginning of the file?
> 
> ```
> from tensorflow.keras.mixed_precision import experimental as mixed_precision
> policy = mixed_precision.Policy('mixed_float16')
> mixed_precision.set_policy(policy)
> ```



> @ion-elgreco Can you provide the full code together with batchsize etc hyper parameters, train & val datasets etc. Otherwise, it's a bit too much effort to reproduce.
> 
> Does it also crash after adding this to the beginning of the file?
> 
> ```
> from tensorflow.keras.mixed_precision import experimental as mixed_precision
> policy = mixed_precision.Policy('mixed_float16')
> mixed_precision.set_policy(policy)
> ```

This is the code: https://github.com/ion-elgreco/thesis-FER/blob/main/models.py

However, I cannot share the datasets because of a legal agreement. The thing is, I hadn't changed a bit to this code. It worked fine before on my i7 7700k + rtx 3080, but after upgrading to a 5800x processor and having to reinstall windows, I keep getting this issue. I even tried reinstalling windows, 5 times..

**Comment by ahtik** [comment]
@ion-elgreco I'm not aware of a good reason how the training could become slower after enabling mixed precision...

Do I understand correctly that you're also running TF2.3 right now inside this new desktop with a Ryzen CPU, and just switching between active CUDA version? So the TF2.3 is not crashing and TF2.4 does with the **exact** same hw and software conf.

Couldn't rule out the HW issue as well, this one is from pytorch, but still relevant. https://discuss.pytorch.org/t/runtimeerror-cudnn-status-internal-error-when-l-run-the-program-for-a-second-time/2960/4

As there can be many causes for CUDNN_STATUS_INTERNAL_ERROR, the only pragmatic way to help here, is to also provide the dataset (processed enough to not cause the legal issues), that would be as small and compact as possible to reproduce the issue with TF2.4. Also include the affected version info etc. Might be worth opening as a separate issue, as the cause seems to be different from the one from @nectario.

I noticed a comment of using conda. To be 100% clean, I'd definitely try to reproduce this also with a clean venv install with pip. Smth like
```
cd projdir
\python38\python.exe -m venv .venv
.\venv\Scripts\activate or .\venv\Scripts\Activate.ps1 for powershell
pip install tensorflow
python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])));print(tf.config.list_physical_devices('GPU'))"
python yourtrainging.py
```

**Comment by ion-elgreco** [comment]
> @ion-elgreco I'm not aware of a good reason how the training could become slower after enabling mixed precision... Initial graph building, yes.
> 
> Do I understand correctly that you're also running TF2.3 right now inside this new desktop with a Ryzen CPU, and just switching between active CUDA version? So the TF2.3 is not crashing and TF2.4 does with the **exact** same hw and software conf.
> 
> Couldn't rule out the HW issue as well, this one is from pytorch, but still relevant. https://discuss.pytorch.org/t/runtimeerror-cudnn-status-internal-error-when-l-run-the-program-for-a-second-time/2960/4
> 
> As there can be many causes for CUDNN_STATUS_INTERNAL_ERROR, the only pragmatic way to help here, is to also provide the dataset (processed enough to not cause the legal issues), that would be as small and compact as possible to reproduce the issue with TF2.4. Also include the affected version info etc. Might be worth opening as a separate issue, as the cause seems to be different from the one from @nectario.
> 
> I noticed a comment of using conda. To be 100% clean, I'd definitely try to reproduce this also with a clean venv install with pip. Smth like
> 
> ```
> cd projdir
> \python38\python.exe -m venv .venv
> .\venv\Scripts\activate or .\venv\Scripts\Activate.ps1 for powershell
> pip install -U pip
> pip install tensorflow
> python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])));print(tf.config.list_physical_devices('GPU'))"
> python yourtrainging.py
> ```

I have two environments, my main one with TF 2.4 with Cuda 11 and Cudnn 8.0.5, and TF 2.3 with Cuda 10.1 with Cudnn 7.6.5. After running TF 2.3 longer it also seems to crash so I was mistaken that it only was with Cuda 11+ and Cudnn 8+.

I am not only getting CUDNN_STATUS_INTERNAL_ERROR, I also get:

```
Failed to call ThenRnnBackward with model config: [rnn_mode, rnn_input_mode, rnn_direction_mode]: 2, 0, 0 , [num_layers, input_size, num_units, dir_count, max_seq_length, batch_size, cell_num_units]: [1, 4608, 100, 1, 60, 32, 100] 
	 [[{{node gradients/CudnnRNN_grad/CudnnRNNBackprop}}]]
	 [[PartitionedCall]] [Op:__inference_train_function_1127558]
```

And some other cudnn crashes, they all happen at random and the error code is each time random.

I'll try in a clean venv environment later today.

**Comment by ion-elgreco** [comment]
> @ion-elgreco I'm not aware of a good reason how the training could become slower after enabling mixed precision... Initial graph building, yes.
> 
> Do I understand correctly that you're also running TF2.3 right now inside this new desktop with a Ryzen CPU, and just switching between active CUDA version? So the TF2.3 is not crashing and TF2.4 does with the **exact** same hw and software conf.
> 
> Couldn't rule out the HW issue as well, this one is from pytorch, but still relevant. https://discuss.pytorch.org/t/runtimeerror-cudnn-status-internal-error-when-l-run-the-program-for-a-second-time/2960/4
> 
> As there can be many causes for CUDNN_STATUS_INTERNAL_ERROR, the only pragmatic way to help here, is to also provide the dataset (processed enough to not cause the legal issues), that would be as small and compact as possible to reproduce the issue with TF2.4. Also include the affected version info etc. Might be worth opening as a separate issue, as the cause seems to be different from the one from @nectario.
> 
> I noticed a comment of using conda. To be 100% clean, I'd definitely try to reproduce this also with a clean venv install with pip. Smth like
> 
> ```
> cd projdir
> \python38\python.exe -m venv .venv
> .\venv\Scripts\activate or .\venv\Scripts\Activate.ps1 for powershell
> pip install -U pip
> pip install tensorflow
> python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])));print(tf.config.list_physical_devices('GPU'))"
> python yourtrainging.py
> ```

That is interesting. It could be definitely the case that the 5800x is causing it..

**Comment by ahtik** [comment]
@ion-elgreco It could also help if you have a non-Ryzen GPU setup somewhere that can be used for this private dataset training making sure it works there with both TF2.3 and TF2.4.

**Comment by ion-elgreco** [comment]
> @ion-elgreco It could also help if you have a non-Ryzen GPU setup somewhere that can be used for this private dataset training making sure it works there with both TF2.3 and TF2.4.

I had an i7-7700k a week ago, with TF2.3 and TF2.4 my code worked. I cant switch back to Intel because then I would have to lend components from someone.

**Comment by nectario** [comment]
One interesting thing that is happening right now in regards to this issue:
After installing TensorFlow 2.4.0, and ugrading my CUDA to 11.0 and 8.02 then downgrading to TensorFlow 2.3.1 and downgrading my CUDA to the version supported by TensorFlow, the issue seems to have stopped happening.

I have trained twice already with 1500 epochs, and no failures so far....

**Comment by nectario** [comment]
My NVIDIA driver is also different. I am using the studio version rather than the gaming version.

**Comment by nectario** [comment]
It is a bit slower. I remember running this at 1sec per epoch.

**Comment by ion-elgreco** [comment]
> It is a bit slower. I remember running this at 1sec per epoch.

Studio and game ready drivers are the same if the versions are the same

**Comment by nectario** [comment]
> > It is a bit slower. I remember running this at 1sec per epoch.
> 
> Studio and game ready drivers are the same if the versions are the same

I see. It's strange, issue not happening now.

**Comment by ion-elgreco** [comment]
> > > It is a bit slower. I remember running this at 1sec per epoch.
> > 
> > 
> > Studio and game ready drivers are the same if the versions are the same
> 
> I see. It's strange, issue not happening now.

I have tried either version of studio and game-ready. Tried all of them which are available for the RTX 3080, but the issue persists.

**Comment by nectario** [comment]
This is my training output wen it starts:

`2020-12-28 13:59:46.246094: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library nvcuda.dll
2020-12-28 13:59:46.275929: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:83:00.0 name: TITAN V computeCapability: 7.0
coreClock: 1.455GHz coreCount: 80 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 607.97GiB/s
2020-12-28 13:59:46.276253: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudart64_101.dll
2020-12-28 13:59:46.288071: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cublas64_10.dll
2020-12-28 13:59:46.294217: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cufft64_10.dll
2020-12-28 13:59:46.296723: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library curand64_10.dll
2020-12-28 13:59:46.305299: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusolver64_10.dll
2020-12-28 13:59:46.311567: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusparse64_10.dll
2020-12-28 13:59:46.346196: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudnn64_7.dll
2020-12-28 13:59:46.346475: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0
2020-12-28 13:59:46.362025: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20e513b8d80 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-12-28 13:59:46.362291: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-12-28 13:59:46.362731: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:83:00.0 name: TITAN V computeCapability: 7.0
coreClock: 1.455GHz coreCount: 80 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 607.97GiB/s
2020-12-28 13:59:46.363085: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudart64_101.dll
2020-12-28 13:59:46.363276: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cublas64_10.dll
2020-12-28 13:59:46.363465: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cufft64_10.dll
2020-12-28 13:59:46.363648: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library curand64_10.dll
2020-12-28 13:59:46.363841: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusolver64_10.dll
2020-12-28 13:59:46.364033: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusparse64_10.dll
2020-12-28 13:59:46.364218: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudnn64_7.dll
2020-12-28 13:59:46.364439: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0
2020-12-28 13:59:47.265842: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-12-28 13:59:47.266092: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]      0 
2020-12-28 13:59:47.266217: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0:   N 
2020-12-28 13:59:47.266667: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10245 MB memory) -> physical GPU (device: 0, name: TITAN V, pci bus id: 0000:83:00.0, compute capability: 7.0)
2020-12-28 13:59:47.271449: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20ea8d5c4c0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-12-28 13:59:47.271684: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): TITAN V, Compute Capability 7.0`

**Comment by ion-elgreco** [comment]
> This is my training output wen it starts:
> 
> `2020-12-28 13:59:46.246094: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library nvcuda.dll 2020-12-28 13:59:46.275929: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: pciBusID: 0000:83:00.0 name: TITAN V computeCapability: 7.0 coreClock: 1.455GHz coreCount: 80 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 607.97GiB/s 2020-12-28 13:59:46.276253: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudart64_101.dll 2020-12-28 13:59:46.288071: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cublas64_10.dll 2020-12-28 13:59:46.294217: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cufft64_10.dll 2020-12-28 13:59:46.296723: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library curand64_10.dll 2020-12-28 13:59:46.305299: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusolver64_10.dll 2020-12-28 13:59:46.311567: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusparse64_10.dll 2020-12-28 13:59:46.346196: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudnn64_7.dll 2020-12-28 13:59:46.346475: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0 2020-12-28 13:59:46.362025: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20e513b8d80 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2020-12-28 13:59:46.362291: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2020-12-28 13:59:46.362731: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: pciBusID: 0000:83:00.0 name: TITAN V computeCapability: 7.0 coreClock: 1.455GHz coreCount: 80 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 607.97GiB/s 2020-12-28 13:59:46.363085: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudart64_101.dll 2020-12-28 13:59:46.363276: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cublas64_10.dll 2020-12-28 13:59:46.363465: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cufft64_10.dll 2020-12-28 13:59:46.363648: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library curand64_10.dll 2020-12-28 13:59:46.363841: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusolver64_10.dll 2020-12-28 13:59:46.364033: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cusparse64_10.dll 2020-12-28 13:59:46.364218: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library cudnn64_7.dll 2020-12-28 13:59:46.364439: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0 2020-12-28 13:59:47.265842: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor with strength 1 edge matrix: 2020-12-28 13:59:47.266092: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263] 0 2020-12-28 13:59:47.266217: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0: N 2020-12-28 13:59:47.266667: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10245 MB memory) -> physical GPU (device: 0, name: TITAN V, pci bus id: 0000:83:00.0, compute capability: 7.0) 2020-12-28 13:59:47.271449: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20ea8d5c4c0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices: 2020-12-28 13:59:47.271684: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): TITAN V, Compute Capability 7.0`

So it's working now with the exact same model and data? Because somehow I can run LSTM fine with some small text data set, but not my big image dataset.

**Comment by nectario** [comment]
> > This is my training output wen it starts:
> 
> So it's working now with the exact same model and data? Because somehow I can run LSTM fine with some small text data set, but not with my big image dataset.

It's the same large data although slightly different (some content is different).

**Comment by nectario** [comment]
The difference right now is that I can control training based to cutoff dates. My cutoff date right now is at Sept 30th.

**Comment by nectario** [comment]
The main difference is the NVIDIA driver. It was different before. I am now using: 460.89 (Studio Version)

**Comment by nectario** [comment]
My existing training would almost always fail. I am on the third run right now.

**Comment by ion-elgreco** [comment]
I found the cause of cuda crashes... it is the RMSprop optimizer. With all other leaners, I can run the LSTM fine, this is mad...

**Comment by nectario** [comment]
Wow! What about adam? That is what I have.

**Comment by ion-elgreco** [comment]
> Wow! What about adam? That is what I have.

Oh never mind.. With Adam it will crash simply not as fast as with RMSprop.

**Comment by ion-elgreco** [comment]
Ok I have some good news :) I was messing around with WSL2 to simply run tensorflow-gpu in Linux. So I had to install the cuda capable driver 465.12 (developer) driver, which now does not cause anymore crashes in Windows.

PS
In case someone has experience with WSL2, I am struggling to get it to recognize my GPU.

**Comment by nectario** [comment]
Wow, awesome! So it's a driver issue! 

Funny you are messing with WSL2 -- I was doing the same a couple of days back. Haven't gotten too deep into it. Maybe this will help:
[https://ubuntu.com/blog/getting-started-with-cuda-on-ubuntu-on-wsl-2](url)

**Comment by ion-elgreco** [comment]
> Wow, awesome! So it's a driver issue!
> 
> Funny you are messing with WSL2 -- I was doing the same a couple of days back. Haven't gotten too deep into it. Maybe this will help:
> https://ubuntu.com/blog/getting-started-with-cuda-on-ubuntu-on-wsl-2

Yeah I found that guide already, but Ubuntu with WSL2 won't recognize my gpu. Even though  I have installed the right driver.

**Comment by nectario** [comment]
Same for me. 2.4 was core dumping and I had to install 2.3.1 but it was CPU version. If I get it to work with the GPU I will let you know.  

But man, after all these months, finally a solution with the above problem! Today is the first time I can train with no issue for hours!

**Comment by ion-elgreco** [comment]
> Same for me. 2.4 was core dumping and I had to install 2.3.1 but it was CPU version. If I get it to work with the GPU I will let you know.
> 
> But man, after all these months, finally a solution with the above problem! Today is the first time I can train with no issue for hours!

Yeah it's nice to see TF working properly again! If you got GPU working in WSL Lett me know :)  I don't know where I read it but it seems like they want to push GPU support more towards WSL.

**Comment by nectario** [comment]
You got it! That would be great.

**Comment by ion-elgreco** [comment]
> You got it! That would be great.

I think I have found the cause. I am not on a windows insider build. I am on:

```
Edition	Windows 10 Pro
Version	20H2
Installed on	‎20/‎12/‎2020
OS build	19042.685
Experience	Windows Feature Experience Pack 120.2212.551.0
```

But according to the cuda docs I should be on **20145 or higher** and for tensorflow directml 20150 or higher..

**Comment by nectario** [comment]
Ah, great. I just initiated the enrollment on the Windows Insider Build. This is great.

**Comment by beelzmon** [comment]
I know you found a fix, but another fix i found is changing the settings so they do not align with the cudnn requirements:
https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM i.e. adding recurrent dropout - which can be very useful.
If you change these settings it forces it to use the generic GPU kernel and seems to work even if at a performance cost.

**Comment by ion-elgreco** [comment]
> I know you found a fix, but another fix i found is changing the settings so they do not align with the cudnn requirements:
> https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM i.e. adding recurrent dropout - which can be very useful.
> If you change these settings it forces it to use the generic GPU kernel and seems to work even if at a performance cost.

That is not a fix though, because you are not using the Cudnn kernel. It's a nice trick though, I used it too in the mean while by disabling eager execution

**Comment by nectario** [comment]
Sorry, this solution will not work for me. I need the speed increase that the CUDNN optimization provides due to the large number of epochs.

**Comment by beelzmon** [comment]
Have you tried and compared the speed? From what you said you are using a lot of RNN -- my trick doesn't hurt performance too much for one RNN layer as the rest of the layers (CNN, Dense) still use CUDNN. With many RNN layers maybe not. Another option is to use ConvLSTM2D with appropriately resized inputs. That also worked for me.

**Comment by nectario** [comment]
> > You got it! That would be great.
> 
> I think I have found the cause. I am not on a windows insider build. I am on:
> 
> ```
> Edition	Windows 10 Pro
> Version	20H2
> Installed on	‎20/‎12/‎2020
> OS build	19042.685
> Experience	Windows Feature Experience Pack 120.2212.551.0
> ```
> 
> But according to the cuda docs I should be on **20145 or higher** and for tensorflow directml 20150 or higher to be able to use cuda in wsl2.

@ion-elgreco 

I finally got TensorFlow with CUDA/CUDNN to work WSL 2 (without using Docker containers). Just a few things:

1. I had to use TF 2.3.1 as 2.4 would core dump.
2. It's actually slower: It takes 6 seconds per epoch versus 2 seconds on Windows 10.

If the speed issue can be resolved hands down this would be the only environment I will use!

**Comment by ion-elgreco** [comment]
> 1. core dump.

Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?

**Comment by nectario** [comment]
> > 1. core dump.
> 
> Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?

Here is the info:

**Build:** 10.0.21277 Build 21277
**Nvidia Driver Version:** 460.89 (Studio)

As for the commands, I ran a bunch but I think these made it work:

```
wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb

sudo apt install ./nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb

sudo apt install nvidia-cuda-toolkit

download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):

https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb

sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb


sudo apt install nvidia-driver-440

type nvidia-smi to verify

Make sure you install tensorflow 3.1
```

**Comment by nectario** [comment]
@ion-elgreco 
I was trying many things for a long time not realizing the issue was with TF 2.4.0

**Comment by ion-elgreco** [comment]
> > > 1. core dump.
> > 
> > 
> > Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?
> 
> Here is the info:
> 
> **Build:** 10.0.21277 Build 21277
> **Nvidia Driver Version:** 460.89 (Studio)
> 
> As for the commands, I ran a bunch but I think these made it work:
> 
> ```
> wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> 
> sudo apt install ./nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> 
> sudo apt install nvidia-cuda-toolkit
> 
> download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):
> 
> https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> 
> sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> 
> 
> sudo apt install nvidia-driver-440
> 
> type nvidia-smi to verify
> 
> Make sure you install tensorflow 3.1
> ```

Hmm interesting, because you are installing the Nvidia driver inside the WSL ubuntu env. According to env to properly use your GPU is to install the driver only on your main OS and not inside the WSL ubuntu. Maybe that's why it's slower?

I'll give your commands a try later this week! Thanks.

**Comment by nectario** [comment]
Maybe you are right. I was having a ton of issues before that. I will try uninstalling it and see what happens.

**Comment by ion-elgreco** [comment]
> @ion-elgreco
> 
> > > > > 1. core dump.
> > > > 
> > > > 
> > > > Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?
> > > 
> > > 
> > > Here is the info:
> > > **Build:** 10.0.21277 Build 21277
> > > **Nvidia Driver Version:** 460.89 (Studio)
> > > As for the commands, I ran a bunch but I think these made it work:
> > > ```
> > > wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > 
> > > sudo apt install ./nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > 
> > > sudo apt install nvidia-cuda-toolkit
> > > 
> > > download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):
> > > 
> > > https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > 
> > > sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > 
> > > 
> > > sudo apt install nvidia-driver-440
> > > 
> > > type nvidia-smi to verify
> > > 
> > > Make sure you install tensorflow 3.1
> > > ```
> > 
> > 
> > Hmm interesting, because you are installing the Nvidia driver inside the WSL ubuntu env. According to env to properly use your GPU is to install the driver only on your main OS and not inside the WSL ubuntu. Maybe that's why it's slower?
> > I'll give your commands a try later this week! Thanks.
> 
> Maybe you are right. I was having a ton of issues before that. I will try uninstalling it and see what happens.

Only the developer driver version is supported according to Nvidia. https://developer.nvidia.com/cuda/wsl/download

**Comment by nectario** [comment]
> > @ion-elgreco
> > > > > > 1. core dump.
> > > > > 
> > > > > 
> > > > > Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?
> > > > 
> > > > 
> > > > Here is the info:
> > > > **Build:** 10.0.21277 Build 21277
> > > > **Nvidia Driver Version:** 460.89 (Studio)
> > > > As for the commands, I ran a bunch but I think these made it work:
> > > > ```
> > > > wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > > 
> > > > sudo apt install ./nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > > 
> > > > sudo apt install nvidia-cuda-toolkit
> > > > 
> > > > download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):
> > > > 
> > > > https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > > 
> > > > sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > > 
> > > > 
> > > > sudo apt install nvidia-driver-440
> > > > 
> > > > type nvidia-smi to verify
> > > > 
> > > > Make sure you install tensorflow 3.1
> > > > ```
> > > 
> > > 
> > > Hmm interesting, because you are installing the Nvidia driver inside the WSL ubuntu env. According to env to properly use your GPU is to install the driver only on your main OS and not inside the WSL ubuntu. Maybe that's why it's slower?
> > > I'll give your commands a try later this week! Thanks.
> > 
> > 
> > Maybe you are right. I was having a ton of issues before that. I will try uninstalling it and see what happens.
> 
> Only the developer driver version is supported according to Nvidia. https://developer.nvidia.com/cuda/wsl/download

Actually, I had the correct driver version installed (the one in the link you sent). I confused the new driver update message with the version installed. I am still at 6 seconds even after uninstalling nvidia-440...

**Comment by ion-elgreco** [comment]
> > > @ion-elgreco
> > > > > > > 1. core dump.
> > > > > > 
> > > > > > 
> > > > > > Could you maybe explain in detail, which OS build, driver versions etc you needed to get it to work?
> > > > > 
> > > > > 
> > > > > Here is the info:
> > > > > **Build:** 10.0.21277 Build 21277
> > > > > **Nvidia Driver Version:** 460.89 (Studio)
> > > > > As for the commands, I ran a bunch but I think these made it work:
> > > > > ```
> > > > > wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > > > 
> > > > > sudo apt install ./nvidia-machine-learning-repo-ubuntu2004_1.0.0-1_amd64.deb
> > > > > 
> > > > > sudo apt install nvidia-cuda-toolkit
> > > > > 
> > > > > download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):
> > > > > 
> > > > > https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > > > 
> > > > > sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb
> > > > > 
> > > > > 
> > > > > sudo apt install nvidia-driver-440
> > > > > 
> > > > > type nvidia-smi to verify
> > > > > 
> > > > > Make sure you install tensorflow 3.1
> > > > > ```
> > > > 
> > > > 
> > > > Hmm interesting, because you are installing the Nvidia driver inside the WSL ubuntu env. According to env to properly use your GPU is to install the driver only on your main OS and not inside the WSL ubuntu. Maybe that's why it's slower?
> > > > I'll give your commands a try later this week! Thanks.
> > > 
> > > 
> > > Maybe you are right. I was having a ton of issues before that. I will try uninstalling it and see what happens.
> > 
> > 
> > Only the developer driver version is supported according to Nvidia. https://developer.nvidia.com/cuda/wsl/download
> 
> Actually, I had the correct driver version installed (the one in the link you sent). I confused the new driver update message with the version installed. I am still at 6 seconds even after uninstalling nvidia-440...

It's probably the slow IO with WSL. Do you need to load a lot of files?

**Comment by ahtik** [comment]
@nectario There's an interesting observation and a fix regarding the TDR setting causing issues, MAYBE it's also relevant to your CUDA_ERROR_LAUNCH_FAILED case.

**Comment by nectario** [comment]
> @nectario There's an interesting observation and a fix regarding the TDR setting causing issues, MAYBE it's also relevant to your CUDA_ERROR_LAUNCH_FAILED case. [#45987 (comment)](https://github.com/tensorflow/tensorflow/issues/45987#issuecomment-774401485)

@ahtik Thank you for bringing this to our attention! I just applied the fix. Let's see...!

**Comment by ericvoots** [comment]
were you able to resolve the error?  I am getting the same thing currently.

**Comment by nectario** [comment]
@ericvoots Yes. It has not happened ever since I upgraded my driver. I also applied the above fix. Although this issue was already resolved with the driver upgrade.

**Comment by pawarrick** [comment]
@ion-elgreco, upgrading to the driver version 465.12 (developer)  solved this problem for me as well with TF2.2/Cuda 10.1.  Wonder whether this version in off the master release path and when a supported release driver will work.  I did not try the latest suggested driver to check.  Just happy I'm running again. Many thanks

**Comment by i-chaochen** [comment]
> @ericvoots Yes. It has not happened ever since I upgraded my driver. I also applied the above fix. Although this issue was already resolved with the driver upgrade.

Do you mean upgrade the drive version not the cuda version? I think this Bcuda10.1_amd64.deb link is not working and make sure tensorflow 3.1 is installed is not correct.

```
download cudnn 7.65 for ubuntu 18.04 (there is no 20.04 version and you need to login and manually download it):

https://developer.nvidia.com/compute/machine-learning/cudnn/secure/7.6.5.32/Production/10.1_20191031/Ubuntu18_04-x64/libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb

sudo apt install ./libcudnn7_7.6.5.32-1%2Bcuda10.1_amd64.deb


sudo apt install nvidia-driver-440

type nvidia-smi to verify

Make sure you install tensorflow 3.1

```

I have the same problem, I think it's because the versions of cuda and cudnn have been messed up.

My tensorflow is 2.4.1
CUDA Version: 11.2
Driver Version: 460.39

I installed cudnn by ```sudo dpkg -i libcudnn8_8.1.0.77-1+cuda11.2_amd64.deb``` and ```sudo dpkg -i libcudnn8-dev_8.1.0.77-1+cuda11.2_amd64.deb```

But I cannot find the correct output via
```
cat /usr/include/cudnn.h | grep CUDNN_MAJOR -A 2
```

**Comment by nectario** [comment]
Is this issue addressed with the 2.7 release?

**Comment by nightpizza123** [comment]
sorry everyone, I don't know which comment should be responded. I find I upgraded requests using sudo pip install --upgrade requests and now I can't import it anymore.

**Comment by gadagashwini** [comment]
HI @nectario, This issue is fixed in Tf v2.9. with Driver Version: 470.141.03  and  CUDA Version: 11.4 on Tesla T4. Could you test it on latest Tensorflow version and confirm the same. Thank you!
